libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -947,19 +1005,26 @@ string YulUtilFunctions::arrayDataAreaFunction(ArrayType const& _type)
 
 string YulUtilFunctions::storageArrayIndexAccessFunction(ArrayType const& _type)
 {
-	solUnimplementedAssert(_type.baseType()->storageBytes() > 16, "");
-
 	string functionName = "storage_array_index_access_" + _type.identifier();
 	return m_functionCollector.createFunction(functionName, [&]() {
 		return Whiskers(R"(
 			function <functionName>(array, index) -> slot, offset {
-				if iszero(lt(index, <arrayLen>(array))) {
+				let arrayLength := <arrayLen>(array)
+				if iszero(lt(index, arrayLength)) {
 					invalid()
 				}
 
 				let data := <dataAreaFunc>(array)
 				<?multipleItemsPerSlot>
-
+					let itemsPerSlot := div(0x20, <storageBytes>)
+					slot := add(data, div(index, itemsPerSlot))
+					offset := mod(index, itemsPerSlot)
+					<?isBytesArray>
+						if lt(arrayLength, 0x20) {
```

Comment:
In case this is true, we needlessly computed `let data := <dataAreaFunc>(array)` - is there a way to do it nicely so this is only done for long arrays?

Reply:
There must be a way, will figure it out

Reply:
I did change it a bit. It has some duplicated code, but there is no needless computations now